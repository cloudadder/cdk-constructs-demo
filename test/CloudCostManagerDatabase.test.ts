import { App } from 'aws-cdk-lib';
import { Annotations, Template } from 'aws-cdk-lib/assertions';
import { TestStackWithDatabaseNegitive, TestStackWithDatabasePositive } from './TestStacks';

describe('Cloud Cost Manager Database', () => {
  test('(DatabaseTestNegitive): confirm cloud cost management reject stack database', () => {
    const app = new App();
    var stack = new TestStackWithDatabaseNegitive(app, 'TestStackWithDatabase');

    Annotations.fromStack(stack).hasError('/TestStackWithDatabase',
      'Do not use MSSQL Enterprise Edition, it is too expensive.Multi-AZ is not supported in Non Production Environments.',
    );
  });

  test('(DatabaseTestPositive): confirm cloud cost management is implemented for database', () => {
    const app = new App();
    var stack = new TestStackWithDatabasePositive(app, 'TestStackWithDatabase');
    Annotations.fromStack(stack).hasInfo('/TestStackWithDatabase',
      'CloudCostManager validation passed',
    );

    Template.fromStack(stack).hasResourceProperties('AWS::RDS::DBInstance', {
      Tags: [
        {
          Key: 'cloud-cost-manager:check:database',
          Value: 'pass',
        },
        {
          Key: 'cloud-cost-manager:customer-name',
          Value: 'acme-co',
        },
        {
          Key: 'cloud-cost-manager:env-name',
          Value: 'production',
        },
      ],
    });
  });

  test('Snapshot Database', () => {
    const app = new App();
    const stack = new TestStackWithDatabasePositive(app, 'test');

    expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
  });
});