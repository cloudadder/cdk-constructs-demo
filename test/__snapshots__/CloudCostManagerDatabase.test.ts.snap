// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Cost Manager Database Snapshot Database 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DatabaseInstance24D16791": Object {
      "DeletionPolicy": "Snapshot",
      "Properties": Object {
        "AllocatedStorage": "100",
        "CopyTagsToSnapshot": true,
        "DBInstanceClass": "db.m5.large",
        "DBSubnetGroupName": Object {
          "Ref": "DatabaseInstanceSubnetGroupA4FFF2A4",
        },
        "Engine": "sqlserver-se",
        "EngineVersion": "14.00.3356.20.v1",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "DatabaseInstanceSecret2C9E2EE4",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "DatabaseInstanceSecret2C9E2EE4",
              },
              ":SecretString:username::}}",
            ],
          ],
        },
        "StorageType": "gp2",
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:check:database",
            "Value": "pass",
          },
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
        ],
        "VPCSecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "DatabaseInstanceSecurityGroup8BDF0112",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Snapshot",
    },
    "DatabaseInstanceSecret2C9E2EE4": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Generated by the CDK for stack: ",
              Object {
                "Ref": "AWS::StackName",
              },
            ],
          ],
        },
        "GenerateSecretString": Object {
          "ExcludeCharacters": " %+~\`#$&*()|[]{}:;<>?!'/@\\"\\\\",
          "GenerateStringKey": "password",
          "PasswordLength": 30,
          "SecretStringTemplate": "{\\"username\\":\\"admin\\"}",
        },
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
        ],
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "DatabaseInstanceSecretAttachmentF5C7BAE8": Object {
      "Properties": Object {
        "SecretId": Object {
          "Ref": "DatabaseInstanceSecret2C9E2EE4",
        },
        "TargetId": Object {
          "Ref": "DatabaseInstance24D16791",
        },
        "TargetType": "AWS::RDS::DBInstance",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "DatabaseInstanceSecurityGroup8BDF0112": Object {
      "Properties": Object {
        "GroupDescription": "Security group for DatabaseInstance database",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DatabaseInstanceSubnetGroupA4FFF2A4": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Subnet group for DatabaseInstance database",
        "SubnetIds": Array [
          Object {
            "Ref": "TestVPCPrivateSubnet1Subnet005D5ABE",
          },
          Object {
            "Ref": "TestVPCPrivateSubnet2Subnet45EC7DE3",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
    "TestVPCD26570D8": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVPCIGW5175BCF9": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVPCPrivateSubnet1DefaultRoute1576D3E9": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "TestVPCPublicSubnet1NATGateway6A40FA74",
        },
        "RouteTableId": Object {
          "Ref": "TestVPCPrivateSubnet1RouteTable1BE239FB",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVPCPrivateSubnet1RouteTable1BE239FB": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVPCPrivateSubnet1RouteTableAssociation3FFF492E": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "TestVPCPrivateSubnet1RouteTable1BE239FB",
        },
        "SubnetId": Object {
          "Ref": "TestVPCPrivateSubnet1Subnet005D5ABE",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVPCPrivateSubnet1Subnet005D5ABE": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVPCPrivateSubnet2DefaultRoute03A159FA": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "TestVPCPublicSubnet2NATGatewayBE12FD22",
        },
        "RouteTableId": Object {
          "Ref": "TestVPCPrivateSubnet2RouteTable9B691DB2",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVPCPrivateSubnet2RouteTable9B691DB2": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVPCPrivateSubnet2RouteTableAssociation08DD625F": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "TestVPCPrivateSubnet2RouteTable9B691DB2",
        },
        "SubnetId": Object {
          "Ref": "TestVPCPrivateSubnet2Subnet45EC7DE3",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVPCPrivateSubnet2Subnet45EC7DE3": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVPCPublicSubnet1DefaultRoute11545FC4": Object {
      "DependsOn": Array [
        "TestVPCVPCGWF063BA4F",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "TestVPCIGW5175BCF9",
        },
        "RouteTableId": Object {
          "Ref": "TestVPCPublicSubnet1RouteTable959CA972",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVPCPublicSubnet1EIPB686BDA0": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "TestVPCPublicSubnet1NATGateway6A40FA74": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "TestVPCPublicSubnet1EIPB686BDA0",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "TestVPCPublicSubnet1SubnetAE333572",
        },
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "TestVPCPublicSubnet1RouteTable959CA972": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVPCPublicSubnet1RouteTableAssociationD7253C7B": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "TestVPCPublicSubnet1RouteTable959CA972",
        },
        "SubnetId": Object {
          "Ref": "TestVPCPublicSubnet1SubnetAE333572",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVPCPublicSubnet1SubnetAE333572": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVPCPublicSubnet2DefaultRouteDCA67168": Object {
      "DependsOn": Array [
        "TestVPCVPCGWF063BA4F",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "TestVPCIGW5175BCF9",
        },
        "RouteTableId": Object {
          "Ref": "TestVPCPublicSubnet2RouteTable9C86C355",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVPCPublicSubnet2EIP59BF172B": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "TestVPCPublicSubnet2NATGatewayBE12FD22": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "TestVPCPublicSubnet2EIP59BF172B",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "TestVPCPublicSubnet2SubnetF2FC6943",
        },
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "TestVPCPublicSubnet2RouteTable9C86C355": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVPCPublicSubnet2RouteTableAssociationBD2B01F3": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "TestVPCPublicSubnet2RouteTable9C86C355",
        },
        "SubnetId": Object {
          "Ref": "TestVPCPublicSubnet2SubnetF2FC6943",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVPCPublicSubnet2SubnetF2FC6943": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "cloud-cost-manager:customer-name",
            "Value": "acme-co",
          },
          Object {
            "Key": "cloud-cost-manager:env-name",
            "Value": "production",
          },
          Object {
            "Key": "Name",
            "Value": "test/TestVPC/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVPCVPCGWF063BA4F": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "TestVPCIGW5175BCF9",
        },
        "VpcId": Object {
          "Ref": "TestVPCD26570D8",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
